ql/src/java/org/apache/hadoop/hive/ql/history/HiveScribeImpl.java

```diff
@@ -0,0 +1,334 @@
+      for (Counters.Group group : counters) {
+        for (Counters.Counter counter : group) {
+          if (first) {
+            first = false;
+          } else {
+            stringBuilder.append(',');
+          }
+          stringBuilder.append(group.getDisplayName());
+          stringBuilder.append('.');
+          stringBuilder.append(counter.getDisplayName());
+          stringBuilder.append(':');
+          stringBuilder.append(counter.getCounter());
+          String tab = getRowCountTableName(counter.getDisplayName());
+          if (tab != null) {
+            if (stringBuilder1.length() > 0) {
+              stringBuilder1.append(",");
+            }
+            stringBuilder1.append(tab);
+            stringBuilder1.append('~');
+            stringBuilder1.append(counter.getCounter());
+            queryInfo.rowCountMap.put(tab, counter.getCounter());
+          }
+        }
+      }
+    } catch (Exception e) {
+      LOG.warn(org.apache.hadoop.util.StringUtils.stringifyException(e));
+    }
+    if (stringBuilder1.length() > 0) {
+      taskInfoMap.get(id).hm.put(HiveHistory.Keys.ROWS_INSERTED.name(), stringBuilder1.toString());
+      queryInfoMap.get(queryId).hm.put(HiveHistory.Keys.ROWS_INSERTED.name(), stringBuilder1
+          .toString());
+    }
+    if (stringBuilder.length() > 0) {
+      taskInfoMap.get(id).hm.put(HiveHistory.Keys.TASK_COUNTERS.name(), stringBuilder.toString());
+    }
+  }
+
+  public static HashMap<String, QueryStats> getQueryStatsMap() {
+    return queryStatsMap;
+  }
+
+  @Override
+  public void printRowCount (String queryId) {
+    return;
+  }
+
+  private void addQueryCompletionTime(String queryId) {
+    Long timeStamp = System.currentTimeMillis();
+
+    QueryStats stats = queryStatsMap.get(queryId);
+    stats.queryEnd = timeStamp;
+  }
+
+  private void addSessionInfo(String queryId) {
+    QueryStats stats = queryStatsMap.get(queryId);
+    SessionState sessionState = SessionState.get();
+
+    if (sessionState != null && sessionState.getUserName() != null
+        && sessionState.getUserIpAddress() != null && sessionState.getSessionId() != null) {
+      stats.username = sessionState.getUserName();
+      stats.IPAddress = sessionState.getUserIpAddress();
+      stats.sessionID = sessionState.getSessionId();
+      stats.database = sessionState.getCurrentDatabase();
+      if (sessionState.getMapRedStats() != null) {
+        stats.mapReduceStatsDesc = sessionState.getMapRedStats().toString();
+      }
+      stats.currentTimeStamp = sessionState.getQueryCurrentTimestamp().toString();
+      stats.mapReduceStats = sessionState.getMapRedStats();
+    }
+  }
+
+  @Override
+  public void endQuery (String queryId) {
+    addQueryCompletionTime(queryId);
+    HiveHistory.QueryInfo queryInfo = queryInfoMap.get(queryId);
+    if (queryInfo == null) {
+      return;
+    }
+
+    addSessionInfo(queryId);
+    scribeQueryMetricEntry(queryId);
+    queryInfoMap.remove(queryId);
```

Comment:
You are removing this `queryId` twice???

Reply:
fixed